components/layout_2020/replaced.rs

```diff
@@ -82,25 +118,83 @@ impl ReplacedContent {
-                block_size.clamp_between_extremums(min_box_size.block, max_box_size.block),
-            )
+        let default_object_size = || {
+            // FIXME:
+            // “If 300px is too wide to fit the device, UAs should use the width of
+            //  the largest rectangle that has a 2:1 ratio and fits the device instead.”
+            // “height of the largest rectangle that has a 2:1 ratio, has a height not greater
+            //  than 150px, and has a width not greater than the device width.”
+            physical::Vec2 {
+                x: Length::new(300.),
+                y: Length::new(150.),
+            }
+            .size_to_flow_relative(mode)
+        };
+        let clamp = |inline_size: Length, block_size: Length| Vec2 {
+            inline: inline_size.clamp_between_extremums(min_box_size.inline, max_box_size.inline),
+            block: block_size.clamp_between_extremums(min_box_size.block, max_box_size.block),
         };
         // https://drafts.csswg.org/css2/visudet.html#min-max-widths
         // https://drafts.csswg.org/css2/visudet.html#min-max-heights
-        let (inline_size, block_size) = match (box_size.inline, box_size.block) {
+        match (box_size.inline, box_size.block) {
             (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::LengthPercentage(block)) => {
                 clamp(inline, block)
             },
             (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::Auto) => {
-                clamp(inline, inline / intrinsic_ratio)
+                let block = if let Some(i_over_b) = intrinsic_ratio {
+                    inline / i_over_b
+                } else if let Some(block) = intrinsic_size.block {
+                    block
+                } else {
+                    default_object_size().block
+                };
+                clamp(inline, block)
             },
             (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(block)) => {
-                clamp(block * intrinsic_ratio, block)
+                let inline = if let Some(i_over_b) = intrinsic_ratio {
+                    block * i_over_b
+                } else if let Some(inline) = intrinsic_size.inline {
+                    inline
+                } else {
+                    default_object_size().inline
+                };
+                clamp(inline, block)
             },
             (LengthOrAuto::Auto, LengthOrAuto::Auto) => {
+                let inline_size =
+                    match (intrinsic_size.inline, intrinsic_size.block, intrinsic_ratio) {
+                        (Some(inline), _, _) => inline,
+                        (None, Some(block), Some(i_over_b)) => {
+                            // “used height” in CSS 2 is always gonna be the intrinsic one,
+                            // since it is available.
+                            block * i_over_b
+                        },
+                        // (Vec2 { inline: None, block: None }, Some(_)) => {
```

Comment:
Commented code that is out-of-date, who would have known? :D Please remove.

Reply:
What do you mean, out of date? The next line is a FIXME comment about a part of the spec that is not implemented yet, and this line has the match arm that we will need to use when we implement that part. I’ll shuffle this so it doesn’t look like comment-out code:

```rust
// FIXME
//
// “If 'height' and 'width' both have computed values of 'auto'
// and the element has an intrinsic ratio but no intrinsic height or width,
// […]”
//
// In this `match` expression this would be an additional arm here:
//
// ```
// (Vec2 { inline: None, block: None }, Some(_)) => {…}
// ```
//
// “[…] then the used value of 'width' is undefined in CSS 2.
// However, it is suggested that, if the containing block's width
// does not itself depend on the replaced element's width,
// then the used value of 'width' is calculated from the constraint
// equation used for block-level, non-replaced elements in normal flow.”